### PR TITLE
added test code

### PR DESCRIPTION
**Pull request by kulrk**
Signed-off-by: kulrk <rohan.1.kulkarni@seagate.com>

# Problem Statement
- Problem statement

# Design
-  For Bug, Describe the fix here.
-  For Feature, Post the link for design

# Coding
   Checklist for Author
-  [ ] Coding conventions are followed and code is consistent

# Testing 
  Checklist for Author
-  [ ] New/Affected tests are executed on Latest Build
-  [ ] Attach test execution logs
-  [ ] Collection tested and no collection error introduced (`pytest --local True --collect-only`)

# Review Checklist 
  Checklist for Author
- [ ] JIRA number/GitHub Issue added to PR
- [ ] PR is self reviewed
- [ ] Jira and state/status is updated and JIRA is updated with PR link
- [ ] Check if the description is clear and explained

# Impact Analysis
  Checklist for Author/Reviewer/GateKeeper
-  [ ] If change in any common function, make sure to update all calls and execute all affected tests.

# Documentation
  Checklist for Author
-  [ ] Changes done to ReadMe / WIKI / Confluence page / Quick Start Guide

### CODE REVIEW

File: tests/dtm/test_single_process_restart.py [hunk 1]
```diff
@@ -168,6 +174,155 @@ def test_write_during_m0d_restart(self):
...
+        self.dtm_obj.perform_write_op(bucket_prefix=bucket_name,
+                                      object_prefix=object_prefix,
+                                      no_of_clients=self.test_cfg['clients'],
+                                      no_of_samples=self.test_cfg['samples'],
+                                      obj_size=self.test_cfg['size'],
+                                      log_file_prefix=log_file_prefix, queue=que)
+        resp = que.get()
+        assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+        self.log.info("Step 1: Start READ Operations :")
```
Comment (priyankaborawake) [inline tests/dtm/test_single_process_restart.py:209]: Correct the step no.

File: tests/dtm/test_single_process_restart.py [hunk 2]
```diff
@@ -168,6 +174,155 @@ def test_write_during_m0d_restart(self):
...
+        proc_read_op = multiprocessing.Process(target=self.dtm_obj.perform_ops,
+                                               args=(workload_info, que,
+                                                     True,
+                                                     True,
+                                                     True, 10))
+        proc_read_op.start()
+        self.log.info("Step 2: Perform Single m0d Process Restart During Write Operations")
```
Comment (priyankaborawake) [inline tests/dtm/test_single_process_restart.py:216]: Correct the operation

File: tests/dtm/test_single_process_restart.py [hunk 3]
```diff
@@ -168,6 +174,155 @@ def test_write_during_m0d_restart(self):
...
+        self.dtm_obj.process_restart(self.master_node_list[0],
+                                     POD_NAME_PREFIX, MOTR_CONTAINER_PREFIX, self.m0d_process)
+
+        self.log.info("Step 3: Check hctl status if all services are online")
+        resp = self.health_obj.is_motr_online()
+        assert_utils.assert_true(resp, 'All services are not online.')
+
+        self.log.info("Step 4: Wait for Write Operation to complete.")
```
Comment (priyankaborawake) [inline tests/dtm/test_single_process_restart.py:224]: Read Operation **

File: tests/dtm/test_single_process_restart.py [hunk 4]
```diff
@@ -168,6 +174,155 @@ def test_write_during_m0d_restart(self):
...
+        if proc_write_op.is_alive():
+            proc_write_op.join()
+        resp = que.get()
+        assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+
+        self.log.info("Step 5: Perform Validate Operations on data written in Step 1:")
```
Comment (priyankaborawake) [inline tests/dtm/test_single_process_restart.py:276]: Validate + delete.

File: tests/dtm/test_single_process_restart.py [hunk 5]
```diff
@@ -168,6 +174,155 @@ def test_write_during_m0d_restart(self):
...
+                                      obj_size=self.test_cfg['size'],
+                                      log_file_prefix=log_file_prefix, queue=que, loop=10)
+        resp = que.get()
+        assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+        self.log.info("Step 2: Start DELETE Operations :")
+        proc_read_op = multiprocessing.Process(target=self.dtm_obj.perform_ops,
+                                               args=(workload_info, que,
+                                                     True,
+                                                     True,
```
Comment (priyankaborawake) [inline tests/dtm/test_single_process_restart.py:310]: validate should be false here.

**Review by priyankaborawake — CHANGES_REQUESTED** [review]
Added minor comments.

File: tests/dtm/test_single_process_restart.py (hunk 2 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py]: Fixed

File: tests/dtm/test_single_process_restart.py (hunk 3 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py]: Done

File: tests/dtm/test_single_process_restart.py [hunk 6]
```diff
@@ -123,6 +123,12 @@ def test_read_during_m0d_restart(self):
             proc_read_op.join()
         resp = que.get()
         assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+
+        self.log.info("Step 6: Perform Delete Operations on data written in Step 1:")
```
Comment (priyankaborawake) [inline tests/dtm/test_single_process_restart.py:128]: Code changes will be added in setup_method to create buckets and hence the delete operation will delete only the objects and not the buckets

Comment (priyankaborawake) [inline tests/dtm/test_single_process_restart.py:128]: Code changes will be added in setup_method to create buckets and hence the delete operation will delete only the objects and not the buckets

**Review by priyankaborawake — APPROVED** [review]

**Comment by priyankaborawake** [comment]
@ApurwaMohite Please review.

File: tests/dtm/test_single_process_restart.py [hunk 7]
```diff
@@ -200,6 +206,12 @@ def test_write_during_m0d_restart(self):
...
+        self.dtm_obj.perform_ops(workload_info, que, True, True, False)
+        resp = que.get()
+        assert_utils.assert_true(resp[0], resp[1])
 
         self.test_completed = True
         self.log.info("ENDED: Verify READ during m0d restart using pkill")
```
Comment (ApurwaMohite) [inline tests/dtm/test_single_process_restart.py:217]: Verify write*

File: tests/dtm/test_single_process_restart.py [hunk 8]
```diff
@@ -200,6 +206,12 @@ def test_write_during_m0d_restart(self):
         self.dtm_obj.perform_ops(workload_info, que, True, True, True)
         resp = que.get()
         assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+
+        self.log.info("Step 6: Perform Delete Operations on data written in Step 1:")
```
Comment (ApurwaMohite) [inline tests/dtm/test_single_process_restart.py:211]: Step 5.

File: tests/dtm/test_single_process_restart.py [hunk 9]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+
+        self.log.info("Step 3 : Perform Single m0d Process Restart During Read Operations")
+        resp = self.dtm_obj.process_restart(master_node=self.master_node_list[0],
+                                            health_obj=self.health_obj,
+                                            pod_prefix=POD_NAME_PREFIX,
+                                            container_prefix=MOTR_CONTAINER_PREFIX,
+                                            process=self.m0d_process, check_proc_state=True)
+        assert_utils.assert_true(resp, "Failure in observed during process restart/recovery")
+
+        self.log.info("Step 5: Wait for READ Operation to complete.")
```
Comment (ApurwaMohite) [inline tests/dtm/test_single_process_restart.py:400]: Please check Step numbers.

**Review by ApurwaMohite — APPROVED** [review]

File: tests/dtm/test_single_process_restart.py [hunk 10]
```diff
@@ -155,6 +155,12 @@ def test_read_during_m0d_restart(self):
             proc_read_op.join()
         resp = que.get()
         assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+
+        self.log.info("Step 6: Perform Delete Operations on data written in Step 1:")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:160]: @priyankaborawake  do we want to shift deletes in teardown if that is not the main intetion of the test? And as discussed we can make delete conditional as to delete only objects or even buckets depending on dtm flag. Will be applicable for all tests where delete is not intention of test
cc: @ApurwaMohite  @kulrk

File: tests/dtm/test_single_process_restart.py [hunk 11]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+        self.dtm_obj.perform_write_op(bucket_prefix=bucket_name,
+                                      object_prefix=object_prefix,
+                                      no_of_clients=self.test_cfg['clients'],
+                                      no_of_samples=self.test_cfg['samples'],
+                                      obj_size=self.test_cfg['size'],
+                                      log_file_prefix=log_file_prefix, queue=que)
+        resp = que.get()
+        assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+        self.log.info("Step 2: Start READ Operations :")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:384]: please add: in background

File: tests/dtm/test_single_process_restart.py [hunk 12]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+                                                     True, 10))
+        proc_read_op.start()
+
+        self.log.info("Step 3 : Perform Single m0d Process Restart During Read Operations")
+        resp = self.dtm_obj.process_restart(master_node=self.master_node_list[0],
+                                            health_obj=self.health_obj,
+                                            pod_prefix=POD_NAME_PREFIX,
+                                            container_prefix=MOTR_CONTAINER_PREFIX,
+                                            process=self.m0d_process, check_proc_state=True)
+        assert_utils.assert_true(resp, "Failure in observed during process restart/recovery")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:398]: "in" not needed here, check wherever applicable

File: tests/dtm/test_single_process_restart.py [hunk 13]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+        proc_write_op = multiprocessing.Process(target=self.dtm_obj.perform_write_op,
+                                                args=(bucket_name, object_prefix,
+                                                      self.test_cfg['clients'],
+                                                      self.test_cfg['samples'],
+                                                      self.test_cfg['size'],
+                                                      log_file_prefix,
+                                                      que, 10))
+        proc_write_op.start()
+
+        self.log.info("Step 3 : Perform Single m0d Process Restart During Read Operations")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:436]: *write Operations

File: tests/dtm/test_single_process_restart.py [hunk 14]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+    @pytest.mark.tags("TEST-41226")
+    def test_continuous_write_during_m0d_restart(self):
+        """Verify continuous WRITE during m0d restart using pkill."""
+        self.log.info("STARTED: Verify continuous WRITE during m0d restart using pkill")
+        bucket_name = 'bucket-test-41226'
+        object_prefix = 'object-test-41226'
+        log_file_prefix = 'test-41226'
+        que = multiprocessing.Queue()
+
+        self.log.info("Step 1: Start write Operations in loop:")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:426]: same as above

File: tests/dtm/test_single_process_restart.py [hunk 15]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+        assert_utils.assert_true(resp, "Failure in observed during process restart/recovery")
+
+        self.log.info("Step 4: Wait for Write Operation to complete.")
+        if proc_write_op.is_alive():
+            proc_write_op.join()
+        resp = que.get()
+        assert_utils.assert_true(resp[0], resp[1])
+        workload_info = resp[1]
+
+        self.log.info("Step 5: Perform Validate and Delete Operations on data written in Step 1:")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:451]: Lets do read and validate here, @priyankaborawake  and we can shift delete to teardown?

File: tests/dtm/test_single_process_restart.py [hunk 16]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+        self.dtm_obj.perform_ops(workload_info, que, True, True, False)
+        resp = que.get()
+        assert_utils.assert_true(resp[0], resp[1])
+
+        self.test_completed = True
+        self.log.info("ENDED: Verify continuous WRITE during m0d restart using pkill")
+
+    @pytest.mark.lc
+    @pytest.mark.dtm
+    @pytest.mark.tags("TEST-41228")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:461]: what's the diff between 41223 and this test?

File: tests/dtm/test_single_process_restart.py [hunk 17]
```diff
@@ -347,3 +359,145 @@ def test_bkt_creation_ios_after_m0d_restart(self):
...
+        workload_info = resp[1]
+        self.log.info("Step 2: Start DELETE Operations :")
+        proc_read_op = multiprocessing.Process(target=self.dtm_obj.perform_ops,
+                                               args=(workload_info, que,
+                                                     True,
+                                                     False,
+                                                     False))
+        proc_read_op.start()
+
+        self.log.info("Step 3 : Perform Single m0d Process Restart During Read Operations")
```
Comment (sonalk0209) [inline tests/dtm/test_single_process_restart.py:488]: same a sabove

**Review by sonalk0209 — COMMENTED** [review]
@kulrk please check comments

File: tests/dtm/test_single_process_restart.py (hunk 8 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py]: fixed

File: tests/dtm/test_single_process_restart.py (hunk 7 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py]: fixed

File: tests/dtm/test_single_process_restart.py (hunk 9 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py]: fixed

File: tests/dtm/test_single_process_restart.py (hunk 11 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py]: added

File: tests/dtm/test_single_process_restart.py (hunk 12 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py:441]: fixed

File: tests/dtm/test_single_process_restart.py (hunk 13 above)
Comment (kulrk) [reply tests/dtm/test_single_process_restart.py]: fixed

File: tests/dtm/test_single_process_restart.py (hunk 15 above)
Comment (priyankaborawake) [reply tests/dtm/test_single_process_restart.py]: @sonalk0209 Deletes for each test can be different , hence not planning to shift delete to teardown.

File: tests/dtm/test_single_process_restart.py (hunk 16 above)
Comment (priyankaborawake) [reply tests/dtm/test_single_process_restart.py:504]: 41223 is for object delete for single bucket. 41228 is object delete for multiple bucket (i.e loop)

File: tests/dtm/test_single_process_restart.py (hunk 15 above)
Comment (sonalk0209) [reply tests/dtm/test_single_process_restart.py]: We can give all the data to one dict which can delete in teardown.

File: tests/dtm/test_single_process_restart.py (hunk 10 above)
Comment (priyankaborawake) [reply tests/dtm/test_single_process_restart.py:160]: Agreed @sonalk0209, 
Separate PR will be raised to move all the delete function for all tests to teardown. 
Except for test where deletes operation are to be main intention of  test.

**Comment by kulrk** [comment]
image


Attaching collection log snapshot

**Review by sonalk0209 — APPROVED** [review]

**Pull request merged**